internal/pkg/manifest/http.go

```diff
@@ -55,9 +49,32 @@ func (r *HTTPOrBool) UnmarshalYAML(value *yaml.Node) error {
 
 // HTTP holds options for application load balancer.
 type HTTP struct {
-	Main                     RoutingRule   `yaml:",inline"`
-	TargetContainerCamelCase *string       `yaml:"targetContainer"` // Deprecated. Maintained for backwards compatibility, use [RoutingRule.TargetContainer] instead.
-	AdditionalRoutingRules   []RoutingRule `yaml:"additional_rules"`
+	Main                     RoutingRule    `yaml:",inline"`
+	TargetContainerCamelCase *string        `yaml:"targetContainer"` // Deprecated. Maintained for backwards compatibility, use [RoutingRule.TargetContainer] instead.
+	GracePeriod              *time.Duration `yaml:"grace_period"`
```

Comment:
> For ALB, grace_period field was under http.healthcheck but grace_period is a service level field, so it should be under http and not http.healthcheck. 

Just in case this wasn't discussed: I recall there was a discussion around this, when `grace_period` was first introduced in https://github.com/aws/copilot-cli/pull/2576: `grace_period` is a health-check related configuration, therefore we put in under `http.health_check`. Although in ECS it is configured as a service attribute, but we tried to model our manifest "in the way that developers think", instead of "in the way how ECS models it".

Reply:
I agree with your point, but it is becoming super weird with our current structure if we keep it inside healthcheck. 
If we keep it inside healthcheck then:
1. Either it has to be under both the main as well as additional routing rules. But it is not a routing rule parameter, so having this parameter in the main or additional rules sounds wrong to me. 
2. Or it has to be under the only Main routing rule, which is also something that is less descriptive of the purpose of that parameter. 

Do let me know your thoughts on this.

Reply:
If multiple containers are handling requests routed from an ELB, I'd assume customers would want to set the grace period to accommodate for the container that is the slowest to start 💭 

An alternative that we can consider is to get the max of the grace periods for individual rules.

```yaml
http:
  path: 'api'
  healthcheck:
     grace_period: 1m
  additional_rules:
      - path: 'user'
        healthcheck:
            path: 'user' 
      - path: 'admin'
        healthcheck:
            path: 'admin'
            grace_period: 10m # The grace period is 10m because 10m > 1m.
```

I see a similar logic in this PR where you get the max of `nlb` and `http`'s `grace_periods`.

Reply:
Yeah that could be another solution, but then we choose max of the grace periods for individual rules. And at the same time if `nlb`'s `grace_period` is also specified then shall we take max of `nlb` and `http` ?

Reply:
I'd imagine we'd want to take the max, if we were to be consistent in the design logic of the "alternative approach".

The pros of the original approach that you purposed by this PR is that it fits nicer with ECS's model. The cons is that it might be too specifically designed for ECS, and misses the abstraction that we've tried to achieve.

For the alternative approach, the pros & cons are the opposite.

I'd also love to hear from the others on this 👍🏼

Reply:
As per the discussion with @iamhopaul123 and @efekarakus, here is the summary of what we are going to do:
1. We agree that [healthcheckgraceperiodseconds](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecs-service.html#cfn-ecs-service-healthcheckgraceperiodseconds) should have been load balancer level setting in ECS. But until they change it, we will keep `grace_period` as is for Main as well as Additional Listener Rules in ALB. That means we will have `grace_period` on `http.Main.HealthCheck` level. So that in the future, whenever ECS changes it, we will open up `grace_period` for additional listener rules as well. For now, we will hide this setting from Additional Rules in our Copilot documents. 
2. In NLB, `grace_period` will be introduced on `nlb.Listener.HealthCheck` level. `grave_period` from additional listeners will be hidden and now shown in our docs. 
3. When ALB and NLB both have `grace_period` set, we will throw a specific error saying they can set either property but not both. Same goes with listener rules and listeners as well. If customers accidentally set `grace_period` on Main as well as Additional Listener Rules, then we error out. 

Do let me know if there are any concerns around this.